### ISSUE TITLE
Documentation on optional settings

### ISSUE THREAD

**Issue by MaxKemman** [senderle/topic-modeling-tool#71]
I am playing around with MALLET and TMT, I see a lot of optional settings with default values, but I have not found explanations for how these settings affect results, and what best practices are for choosing them. I would very much appreciate a guide explaining what each setting does, and advice on how to choose a proper value or evaluate the difference between values.

As an example: why is the default value for number of iterations 400, and how could I decide whether to increase or decrease this value.

**Comment by senderle** [comment]
Here's a first draft. Look it over and let me know if it helps, or if it needs improvement! I'll post a final version after a few people have looked it over.

### Metadata File

A table of comma separated values (CSV) containing metadata to be included in
the output. Each row represents a single document, and each column represents a
metadata feature. The first column **must** be the filename. Many of the most
interesting applications of topic modeling involve analyzing topic distributions
in relation to metadata categories such as date of publication, genre, author,
or publisher.

### Stopword File

A file listing words to exclude from analysis, one word per line. Many users
find that the word lists generated by the tool are easier to interpret when very
common words are excluded (e.g. for English-language models, "a," "the," "of,"
"for," and so on). In some cases, excluding other kinds of words can be
beneficial. For example, topic models of fiction can be skewed by character
names, which may be shared between works that are otherwise unrelated.

Some researchers, especially linguists, argue that excluding common words throws
away important information. For users interested primarily in "content words,"
these exclusions probably won't hurt, and may help. Users interested in
low-level linguistic phenomena should try running the tool without any
exclusions at least once to see how it affects the output.

### Remove default English stopwords

For convenience, the topic modeling tool includes a list of common English
stopwords. It excludes these words by default. Disable this option to prevent
these words from being excluded.

### Preserve case

By default, the topic modeling tool ignores capitalization. 
Generally speaking, typographical and orthographical details such as
capitalization are not relevant to topic models, which pay no attention to word
order, and may be skewed by small inconsistencies. 
However, particular research questions may call for models that are
sensitive to capitalization. Users may sometimes also use capitalization as a
form of annotation or markup that the topic modeling tool should be aware of.
Those users should enable this option.

### Generate HTML output

The topic modeling tool can generate a browsable set of HTML
documents summarizing the results. These are useful for informal inspection and
evaluation, but
users who wish to import ther results into another tool (such as Excel) needn't
leave this option enabled. 

These files can take up a large amount of disk space. Users who are
analyzing very large corpora should plan accordingly, or disable this option.

### Preserve raw MALLET output

Although the CSV output that the topic modeling tool generates can be read by
many tools, some applications require all of the data generated by MALLET, the
topic modeling tool's underlying command-line application. These files can take
up a large amount of disk space, so this option is disabled by default.

### Tokenize with regular expression

Topic modeling uses a "bag of words" model, which means that input texts are
divided up into unordered collections of words before further processing. To
determine where boundaries between words should fall, the topic modeling
tool uses a kind of search string called a regular expression. The default
regular expression here should work well for many languages that use whitespace
to mark boundaries between words. 

Users working with languages that do not separate words with whitespace may need
to preprocess their text using specialized tools. Other users may have texts 
that denote word boundaries in more complex ways. This field allows those 
users to specify alternative word separation schemes.

### Number of iterations

The process that generates a topic model is surprisingly simple. The tool begins
with a random guess about how to divide the corpus into topics. It then makes a
second guess that is slightly more likely to be a good guess; then a third;
then a fourth. Later guesses are more and more likely to be good guesses, but
eventually the gain levels off, and there's no point in continuing after that.

"Iterations" in this context just means "guesses," and this parameter dictates
the total number of guesses to make.

The best value to enter here depends mostly on corpus size. The default, 400
guesses, is probably reasonable for a corpus that contains a few hundred
articles. For a corpus that contains many thousands of articles, or a few
hundred novels, 1000 might be a better number. For very large corpora, even
larger numbers might be appropriate.

Users who are using this tool for more than exploratory purposes should try
a few different values to see how the results change.

### Number of training threads

This dictates the number of computer processes to run simultaneously. If you
have a multicore processor, you can set this to a number that will take full
advantage of all cores. Two processes per core is a common rule of thumb.

If you do not have a multicore processor, the default value, 4, will not make
things faster, but it probably won't make things slower either.

### Number of topic words to print

Every topic in a topic model is associated with a ranked list of words. The list
contains every word in the corpus, but the words at the top of the list are much
more strongly associated with the topic. This determines how far down the list
to go. It's common to print just the first ten or twenty words, but users
sometimes prefer to print more. Some researchers have argued that printing more
is important because words further down the list may provide important
information, or may show when a topic is a combination of two groups of mostly
unrelated words -- a common phenomenon.

### Alpha & Beta optimization frequency

This value controls the way the topic modeling tool adjusts its prior
assumptions about the distributions of words and topics in a corpus. The default
value, 10, tells the tool to update its assumptions once every ten iterations.
Most users will not need to change this value. However, under some
circumstances, users may prefer not to allow the tool to adjust its prior
assumptions at all. Those users should set this value to 0, which disables
adjustments entirely.

For more about these prior assumptions, see below.

### Topic density parameter (Alpha)
### Word density parameter (Beta)

These values are more intimately connected to the underlying algorithm than most
other settings here. They control the kinds of guesses the tool makes about how
topics are associated with documents, and how words are associated with topics.
They do so by controlling how fair these distributions are in the model.

For example, consider a six-sided die. If you roll it many times, and it lands
on all sides about equally often, you'll probably conclude that it's a fair die.
Now suppose you roll a second die many times, and it lands on two sides most of
the time, only rarely landing on any of the other four sides.  You'll probably
conclude that it was biased.

In topic modeling, the Alpha and Beta parameters control how fair or biased the
distributions of topics and words tend to be, according to the model. A high
Alpha value means the model expects documents to have a more even distribution
of topics. A low Alpha value means the model expects documents to have a less
even distribution of topics. Similarly, a high Beta value means the model
expects topics to have a more even distribution of words. A low Beta value means
the model expects topics to have a less even distribution of words.

Changing these values can have dramatic effects, and they aren't always
predictable. In theory, low values might be good for dividing a corpus into
domain-specific vocabularies and specialized document types. High values might
be good for handling overlapping discourses and polyvocal works. But in practice
things may not be so cut and dry.

In short, try to understand what these parameters do, but don't modify them
lightly.

#### PS:

The names of these settings use the word "density" instead of "fairness" or
"evenness." This follows standard usage in statistics. Loosely speaking, a fair
distribution is considered dense because it doesn't have any values that appear
rarely. In this metaphor, rare values are like gaps or holes in the
distribution. The word for a more biased distribution is "sparse." A sparse
distribution has lots of holes.

### Random number generator seed

When the topic modeling tool makes guesses, it does so using a partially random
process. The source of randomness for this process is a sequence of random
numbers. This setting allows you to fix a starting point in that sequence, a
"seed" value. Every time the tool runs with the same seed, it will use the same
sequence of random numbers. If all the other settings remain the same, and the
corpus remains the same, then the output will be identical (in theory). Users
who want their model to be *perfectly* reproducible will want to set this to a
value other than 0, and report the value they used.

### Divide input into n-word chunks

Users who would like to analyze a few very large documents may find that they
get better results by dividing the documents into smaller chunks. This does that
for you automatically.

This strategy may sometimes backfire: the topic modeling tool may determine that
each of the large documents corresponds to one single topic, and assign the
overwhelming majority of each chunk to the topic of the document it came from.
If you have a small number of large documents, this might help improve the
results you get, but it might not.

### Metadata CSV delimiter
### Output CSV delimiter

These settings simply determine how column boundaries are marked in the input
and output files. For example, if your metadata file happens to use tabs instead
of commas, you can enter `\t` as your metadata delimiter, and your file will be
interpreted correctly. Similarly, if you plan to pass the CSV output to a tool
that expects columns to be separated by tabs instead of commas, you can enter
`\t` as your output delimiter, and you won't have to do any conversion.

**Comment by senderle** [comment]
I particularly want expert feedback on this part:

Topic density parameter (Alpha) & Word density parameter (Beta)
These values are more intimately connected to the underlying algorithm than
most other settings here. They control the kinds of guesses the tool makes
about how topics are associated with documents, and how words are associated
with topics. They do so by controlling how fair these distributions are in the
model.

For example, consider a six-sided die. If you roll it many times, and it lands
on all sides about equally often, you'll probably conclude that it's a fair
die. Now suppose you roll a second die many times, and it lands on two sides
most of the time, only rarely landing on any of the other four sides. You'll
probably conclude that it was biased.

In topic modeling, the Alpha and Beta parameters control how fair or biased the
distributions of topics and words tend to be, according to the model. A high
Alpha value means the model expects documents to have a more even distribution
of topics. A low Alpha value means the model expects documents to have a less
even distribution of topics. Similarly, a high Beta value means the model
expects topics to have a more even distribution of words. A low Beta value
means the model expects topics to have a less even distribution of words.

Changing these values can have dramatic effects, and they aren't always
predictable. In theory, low values might be good for dividing a corpus into
domain-specific vocabularies and specialized document types. High values might
be good for handling overlapping discourses and polyvocal works. But in
practice things may not be so cut and dry.

In short, try to understand what these parameters do, but don't modify them
lightly.

PS:
The names of these settings use the word "density" instead of "fairness" or
"evenness." This follows standard usage in statistics. Loosely speaking, a fair
distribution is considered dense because it doesn't have any values that appear
rarely. In this metaphor, rare values are like gaps or holes in the
distribution. The word for a more biased distribution is "sparse." A sparse
distribution has lots of holes.

**Comment by MaxKemman** [comment]
Thank you very much, that really helps my (and I'm sure other's) understanding of what these settings are about, and the extent to which I should change them. One bit that caught my attention:

@mimno writes "The short advice is to set alpha (really the sum of the alpha values for each topic) to 5.0"

But if I look in the TMT settings the alpha is set to 50. Could this be a typo, or is TMT's setting in decimals?

**Comment by senderle** [comment]
50 is still the default in the GUI, and now that I know the recommendations have changed, I will reduce that number! Thanks @mimno for your comments -- I really appreciate it and will make the revisions you suggest.

**Comment by senderle** [comment]
@mimno Another question -- not urgent, answer at your leisure. 

Would you recommend just removing the alpha and beta fields? I added them not too long ago because I found a few cases where adjusting them by hand helped, and I wanted GUI users to have that option. 

But now that I think of it, I was mostly adjusting the 50.0 value downwards, which would be superfluous with the new default. Will GUI users *ever* really want to adjust these?

**Comment by MaxKemman** [comment]
If I may interject, I do appreciate that the alpha and beta fields are visibly present, as it raises awareness that these are settings that affect how topic modelling works. It also provides insight that these settings have be chosen based on 'best practices' rather than as a hard deterministic science, even if I do not intend to change them from the default settings of 0.01 and 5.0

**Comment by senderle** [comment]
@MaxKemman Yes, that's helpful, thanks! I have been uncertain about how much complexity to expose in the GUI, but I think your reasoning in this case is right.

**Comment by MaxKemman** [comment]
I get that, perhaps I'm biased by all the "LDA isn't as simple as it seems" articles so that I want to see the complexities. What might help is distinguishing more between the things that a user should probably change and the things a user probably won't. The way I see it, the 2 settings that a user probably wants to change are "number of iterations" and "number of topic words to print", you could move them to the first pane below "number of topics". This reduces the need to go to the optional settings. 
But for me personally it's not a big issue, thanks to the documentation you've now made available 👍

**Comment by mimno** [comment]
I think it's good to expose the parameter, but *especially* for software aimed a non-experts it's really important to provide guidance about what you should concentrate on. I would explicitly mark each option as one of these categories:

* Always put significant thought into this (num topics, stoplist)
* You may need to modify this in some cases (threads, case folding, alpha)
* Defaults are fine for almost all cases (beta)

**Comment by senderle** [comment]
Closing this, as the requested document [exists](https://senderle.github.io/topic-modeling-tool/documentation/2018/09/27/optional-settings.html), and the changes suggested by dmimno have been separated out into their own issues.
